### ISSUE TITLE
Service not reliably resolving other hostname in the same docker compose network after Docker update v4.21.1

### ISSUE THREAD

**Issue by rija** [docker/for-mac#6929]
### Description

It seems, since update 4.21.1 of Docker Desktop for Mac, a container service that's attached to multiple networks and connecting to another service in one of its network in the same docker-compose file will frequently (but not systematically) fail to resolve the hostname of the target service.

This problem doesn't occur on Mac with 4.21.0 and earlier versions.
This problem doesn't occur if the connecting service has only one network attached to it.

```
psql: error: could not translate host name "server" to address: Name does not resolve
```

### Reproduce

1. Create a `docker-compose.yml` with content:

```
version: '3.7'

services:
  server:
    image: postgres:14.8-alpine3.18
    environment:
      POSTGRES_HOST_AUTH_METHOD: trust
    networks:
      	- db-tier
  client:
    image: postgres:14.8-alpine3.18
    command: bash -c "psql -h server -U postgres -c 'select now();'"
    networks:
      	- db-tier
      	- web-tier

networks:
  web-tier:
  db-tier:
```

2. Create bash script `docker-bug-test.sh` with content: 

```
#!/usr/bin/env bash

# bails on error
set -e

docker-compose up -d server
sleep 1

for i in {0..100}; do docker-compose run --rm client; done

echo "with docker desktop for mac up to 4.21.0, you will see this message. with 4.21.1 you will get an error"
```

3. Run the script

```
$ chmod a+x docker-bug-test.sh
$ ./docker-bug-test.sh
``` 

### Expected behavior

The script doesn't crash and should show the message "with docker desktop for mac up to 4.21.0, you will see this message. with 4.21.1 you will get an error"

### docker version

```bash
Client:
 Cloud integration: v1.0.35
 Version:           24.0.2
 API version:       1.43
 Go version:        go1.20.4
 Git commit:        cb74dfc
 Built:             Thu May 25 21:51:16 2023
 OS/Arch:           darwin/arm64
 Context:           desktop-linux

Server: Docker Desktop 4.21.1 (114176)
 Engine:
  Version:          24.0.2
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.4
  Git commit:       659604f
  Built:            Thu May 25 21:50:59 2023
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.6.21
  GitCommit:        3dce8eb055cbb6872793272b4f20ed16117344f8
 runc:
  Version:          1.1.7
  GitCommit:        v1.1.7-0-g860f061
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client:
 Version:    24.0.2
 Context:    desktop-linux
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.11.0
    Path:     /Users/owner/.docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.19.1
    Path:     /Users/owner/.docker/cli-plugins/docker-compose
  dev: Docker Dev Environments (Docker Inc.)
    Version:  v0.1.0
    Path:     /Users/owner/.docker/cli-plugins/docker-dev
  extension: Manages Docker extensions (Docker Inc.)
    Version:  v0.2.20
    Path:     /Users/owner/.docker/cli-plugins/docker-extension
  init: Creates Docker-related starter files for your project (Docker Inc.)
    Version:  v0.1.0-beta.6
    Path:     /Users/owner/.docker/cli-plugins/docker-init
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc.)
    Version:  0.6.0
    Path:     /Users/owner/.docker/cli-plugins/docker-sbom
  scan: Docker Scan (Docker Inc.)
    Version:  v0.26.0
    Path:     /Users/owner/.docker/cli-plugins/docker-scan
  scout: Command line tool for Docker Scout (Docker Inc.)
    Version:  0.16.1
    Path:     /Users/owner/.docker/cli-plugins/docker-scout

Server:
 Containers: 1
  Running: 1
  Paused: 0
  Stopped: 0
 Images: 1
 Server Version: 24.0.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 3dce8eb055cbb6872793272b4f20ed16117344f8
 runc version: v1.1.7-0-g860f061
 init version: de40ad0
 Security Options:
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 5.15.49-linuxkit-pr
 Operating System: Docker Desktop
 OSType: linux
 Architecture: aarch64
 CPUs: 5
 Total Memory: 7.667GiB
 Name: docker-desktop
 ID: 715bb90b-5f02-4670-9c20-d6ae99811d56
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5555
  127.0.0.0/8
 Live Restore Enabled: false
```

### Diagnostics ID

D347C2EA-DCEC-42DD-9A32-7F83937390DD/20230729174010

### Additional Info

Reproduced on Mac Apple Silicon running Ventura (13.14.1(c)) by two people.

**Comment by Persi** [comment]
Hi @rija, this bug looks like the same problem i described here: https://github.com/docker/for-mac/issues/6918

You could try to upgrade your docker-compose to version 2.20 which should fix this issue. I hope there will be a new Docker for Mac release soon which includes this version...

**Comment by bsousaa** [comment]
Docker Desktop 4.22 is now available. Can you try?

**Comment by rija** [comment]
@Persi : Thanks, I tried a week ago but it was complicated to install docker-compose 2.20 on a Mac without Docker trying to re-install its own version, and after I succeeded, it seemed the error stopped but other issues happened, so decided to wait for official bundled release

@bsousaa: The issue seems resolved with Docker Desktop 4.22, with both the test attached to this issue and the real world project we've detected it. Thanks a lot.

**Comment by ewall-ot** [comment]
I seem to be having this problem again with Docker Desktop for Mac 4.28.0...

**Comment by milas** [comment]
The specific issue here was resolved in Docker Desktop 4.22 (Compose 2.20.2+).

@ewall-ot There was a different networking/DNS issue that could present similarly in 4.28. If you use "Check for updates", 4.29 is now available with Moby v26 that fixes a lot of networking edge cases including this

I'm going to close this now, but if you run into trouble on 4.29, definitely open a new issue!
